### PR TITLE
build(clair): internal clair adapter when install with clair

### PR DESCRIPTION
**Pull request by heww**
1. build support for clair adapter in makefile
2. clair-adapter support in prepare stage
3. create the internal clair adapter when install harbor with clair flag
4. modify test scripts and swagger doc to make the tests match the API changed about scan
5. fix some issues when call scan request to adapter 

Signed-off-by: He Weiwei <hweiwei@vmware.com>

### CODE REVIEW

File: make/photon/prepare/templates/core/env.jinja [hunk 1]
```diff
@@ -36,6 +36,7 @@ CORE_URL={{core_url}}
 CORE_LOCAL_URL={{core_local_url}}
 JOBSERVICE_URL={{jobservice_url}}
 CLAIR_URL={{clair_url}}
+CLAIR_ADAPTER_URL={{clair_adapter_url}}
```
Comment (ywk253100) [inline make/photon/prepare/templates/core/env.jinja]: I think the CLAIR_DB_xxx are all useless?

Comment (ywk253100) [reply make/photon/prepare/templates/core/env.jinja]: The same to `CLAIR_URL `

Comment (heww) [reply make/photon/prepare/templates/core/env.jinja]: clair adapter need the url of the clair and harbor core need the url of the clair adapter

File: make/photon/clair-adapter/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+RUN tdnf install -y sudo >>/dev/null\
+    && tdnf clean all \
+    && mkdir /clair-adapter/ \
+    && groupadd -r -g 10000 clair-adapter \
+    && useradd --no-log-init -m -r -g 10000 -u 10000 clair-adapter
+
+COPY ./make/photon/clair-adapter/binary/harbor-scanner-clair /clair-adapter/clair-adapter
+
+RUN chown -R 10000:10000 /clair-adapter \
```
Comment (ywk253100) [inline make/photon/clair-adapter/Dockerfile]: Should we run this container as non-root user?

Comment (heww) [reply make/photon/clair-adapter/Dockerfile]: Yes

**Comment by steven-zou** [comment]
General comment:

Please fill in meaningful info into the PR description. @heww

**Comment by heww** [comment]
@steven-zou description updated for the pr

File: make/photon/clair-adapter/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    && chmod u+x /clair-adapter/clair-adapter
+
+EXPOSE 8080
+
+HEALTHCHECK --interval=30s --timeout=10s --retries=3 CMD curl -sS 127.0.0.1:8080/healthy || exit 1
```
Comment (steven-zou) [inline make/photon/clair-adapter/Dockerfile]: You should use `/metadata` API endpoint to do check. No healthy endpoint.

File: make/photon/clair-adapter/builder [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+cd `dirname $0`
+cur=$PWD
+
+# the temp folder to store distribution source code...
+TEMP=`mktemp -d ${TMPDIR-/tmp}/clair-adapter.XXXXXX`
+git clone https://github.com/danielpacak/harbor-scanner-clair.git $TEMP
+cd $TEMP; git checkout $VERSION; cd -
+
+echo 'build the clair adapter binary bases on the golang:1.12...'
```
Comment (steven-zou) [inline make/photon/clair-adapter/builder]: no const value defined for golang version?

File: src/core/main.go [hunk 5]
```diff
@@ -146,6 +150,39 @@ func quotaSync() error {
 	return nil
 }
 
+func initClairAdapter(clairAdapterEndpoint string) error {
```
Comment (steven-zou) [inline src/core/main.go]: Move to the `/pkg/scan` package, `main` is not a good place for it. We should keep the `main` as simple as possible.

File: src/pkg/scan/api/scan/base_controller.go [hunk 6]
```diff
@@ -102,6 +102,14 @@ func NewController() Controller {
 	}
 }
 
+func (bc *basicController) jobClient() cj.Client {
```
Comment (steven-zou) [inline src/pkg/scan/api/scan/base_controller.go]: This is ok for now. Just heads up, I'll update it after the rebase.

File: make/photon/clair-adapter/Dockerfile (hunk 3 above)
Comment (heww) [reply make/photon/clair-adapter/Dockerfile]: there is `/healthy` in the clair adapter 
```go
probeRouter.Methods(http.MethodGet).Path("/healthy").HandlerFunc(handler.GetHealthy)
```

`/metadata` maybe need authorization if the adapter implemented the auth.

File: make/photon/clair-adapter/builder (hunk 4 above)
Comment (heww) [reply make/photon/clair-adapter/builder]: changed, the `golang:1.12.5` was used in builder stage.

File: src/core/main.go (hunk 5 above)
Comment (heww) [reply src/core/main.go]: Moved

File: make/photon/clair-adapter/Dockerfile (hunk 3 above)
Comment (steven-zou) [reply make/photon/clair-adapter/Dockerfile]: It is not defined in the open API spec

Comment (heww) [reply make/photon/clair-adapter/Dockerfile]: Yes, it is not in the open API spec, this is the image of clair adapter and `/healthy` is defined by clair adapter i think it is ok to use this for the health checking.

When there is another adapter eg named `foo` maybe its health checking endpoint is `/foo` and then the `HEALTHCHECK` in its docker image maybe `HEALTHCHECK --interval=30s --timeout=10s --retries=3 CMD curl -sS 127.0.0.1:8080/foo || exit 1`

This image care the clair adapter itself, not care the open API spect defined by harbor with other scanners.

Comment (steven-zou) [reply make/photon/clair-adapter/Dockerfile]: Ok, but we need to make sure the `healthy` check endpoint will be always there.

File: src/pkg/scan/init.go [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+	"github.com/goharbor/harbor/src/common/utils/log"
+	"github.com/goharbor/harbor/src/pkg/q"
+	sc "github.com/goharbor/harbor/src/pkg/scan/api/scanner"
+	"github.com/goharbor/harbor/src/pkg/scan/dao/scanner"
+)
+
+// EnsureScanner ensure the scanner which specially name exists in the system
+func EnsureScanner(registration *scanner.Registration) error {
+	q := &q.Query{
+		Keywords: map[string]interface{}{"url": registration.URL},
```
Comment (steven-zou) [inline src/pkg/scan/init.go]: Apppend `PageNumber` and `PageSize` for better performance?

Comment (heww) [reply src/pkg/scan/init.go]: They are not needed here for url in unique in registration table.

**Pull request merged**